### DISCUSSION TITLE
Explorer roadmap planning through Objectives and Key Results

### DISCUSSION THREAD

**Discussion by frol** [near/near-explorer#467] [Ideas]
Hey! Explorer team is working hard on the new features, but we recently put more effort into being more transparent to invite more contributors. You are welcome to contribute your ideas, fixes, and new features!

To be aligned on the direction we are heading, I want to share our vision as to the short-/mid-term plans for Explorer.

The Explorer team collected feedback from our community, internal usage, and even personal pains, and here are the objectives we want to improve in the near future with key results we want to achieve.

### Explorer empowers contract developers building simple contracts

(NOTE: Simple contract are those with up to 5 cross-contract calls which are linear)

To empower the developers we envision the following goals being reached:

* [ ] Explorer represents meta-information about contracts

  *Meta-information about the contracts is the must to reason about a contract: human-readable description, link to the source/homepage, etc.*

* [ ] Explorer enables intuitive access to all the low-level NEAR protocol events (block production, transaction execution, function execution, state changes, accounts and access keys lifetime)

* [ ] Explorer provides a way to interact with contracts right from Explorer UI

  *Calling a view-method right from Explorer UI is the main focus here*

### Explorer is great at onboarding people

We want to make Explorer useful for new-comers, so we expect Explorer to naturally guide users through NEAR protocol concepts, so they can build their mental model. To do so, we think the following goals need to be reached:

* [ ] Navigation is natural and there are no dark corners

  *As an example, we have received quite a few reports that the Nodes Dashboard page was not immediately obvious for how to navigate to*

* [ ] Stats about the chain history with graphs by days/months

### Explorer is useful for validators

Validators are the core team maintaining the network liveness, and to operate they need clear information. Here are the key results we want to achieve:

* [ ] Validators can get all the relevant information about the network status

# Final thoughts

We have already been working on these, but if you feel that we miss something, let us know. We also invite you to contribute to the community

**Comment by eriktrautman** [comment]
I'm not sure if this is covered by the above, but here's a user story that breaks now:

As a multisig user who has submitted a multisig transaction for approval and had it approved, I want to see that transaction in a human-readable way in the explorer like any other transaction.

Currently, the unreadable submitting of the multisig and approval of the transaction both show up but actually what's executed is not discoverable.

**Reply by frol** [reply]
It is a valid concern, but it is a bit more high-level than the Block Explorer wants to be at the moment. We focus on more low-level concepts being fully represented at the moment, but I will record this request to have a special support for multisig. Thank you for bringing it up!

**Reply by luciotato** [reply]
My 2 cents:
* Sounds like you can separate the Explorer in 2 different sub-products

1. The **actual Explorer**, a web interface to the chain, showing all chain data RAW, as it is, even non-human readable data to help **developers** and enthusiasts

2. A **User oriented Dashboard**. A more complex tool that has knowledge of certain contract code (like multisig) which can parse contract-specific logs, can join several transactions in one logical view and show all in a _human readable_, all-in-one-place, logical-view form. It hides by default raw data. It can also interact with some contracts.

The main difference is that **Explorer** is low level and don't have special coding to parse specific contracts logs/output.

OTOH **User Dahsboard** is all about special code to parse specific contracts' output, find related transactions and show digested data to the user.

In my view those are two good and required products, but they're also different beasts.

Stating a clear line and splitting into 2 sub-products will help avoid trying to do a single product that tries to do both things without being good at none (happens a lot). 

Maybe you can extract a low-level common library supporting both subproducts. (it will become the 3rd product which other people can use to build their custom dashboards)

**Reply by frol** [reply]
I agree that those should be two different products, but I still think that we should start with the lower-level till first

**Comment by eriktrautman** [comment]
Highly flexible search which will allow users to enter lots of different data and display it, for example everything that etherscan has basically. A few new things vs today:
+ search by public keys, account hash or lockup
+ sort/filter for search results and pretty much any screen with lots of data (accounts, transactions, whatever)
+ Bonus: export to CSV (not required)

A little more info about validators is nice to have (some of the details on the other tools is good like who is upcoming and who is dropping out, change in stake, etc).

**Reply by frol** [reply]
We will definitely want to add support for these and it falls into the "Explorer is great at onboarding people" objective, but I am afraind Explorer team doesn't have the bandwidth to complete all of them, but certain basic ones fall into the navigation improvements theme I already inclluded in my draft plan :+1:

**Comment by luciotato** [comment]
Incoming transfers are not shown in the account explorer. Only outgoing transfers and contract calls.
If it's not too complex, having incoming transfers listed with the account movements will help a lot.
example:
https://explorer.near.org/accounts/luciotato.near
I received 35N and then 2N in that account but the incoming transfers are not shown.

**Reply by DmitriPro** [reply]
Explorer can have many features. An example is one of the most convenient tools. It's a one-stop station for all ecosystem, can be a wallet, can be a dapp portal, voting tool, transactions review. List of Validators can give more information with links to their websites and show Validators data. Validators can submit all information by themselves also upload a logo. Also can be integrated a feature of DAO, governance voting onchain. For example, can be a poll request for developing a useful tool on NEAR, the community can be engaged to participate in voting by the amount of stake they have. 
Here is an example of multi-feature explorer: 
https://bloks.io/

**Reply by frol** [reply]
@luciotato that is exactly the main focus of Explorer team right now! To achieve it we have implemented a whole new infrastructure for indexing the data :+1:

**Comment by eriktrautman** [comment]
Micro UI improvement -- when a transaction is failed, chane the icon to a big fat X regardless of what type of transaction there is. It's really hard to tell at a glance if a tx failed so it's easy to tell yourself it succeeded since the verbal phrasing of the tx is eg "transferred XN to foo.near.org", which implies success, even though the faint word "failed" is written off to the right.  It should be much more clear in the verbiage and the icon that it failed so people can more easily debug.

**Reply by frol** [reply]
Good point! I created a tracking issue for this https://github.com/near/near-explorer/issues/469
